doc/decisions/1_in_discussion/reference_counting.md

```diff
@@ -0,0 +1,286 @@
+# Reference Counting
+
+## Problem
+
+Elektra's data structures are shared between multiple owners.
+A `Key` can be owned by many different `KeySet`s.
+Both `Key` and `KeySet` can be used in bindings.
+If the languages of those bindings have automatic memory management, this also implies a shared ownership.
+
+To solve this issue of shared ownership, we need reference counting.
+
+An adjacent problem is that the names of `Key`s must be read-only while the `Key` is part of a `KeySet`.
+This can also be solved by a secondary counter just for `KeySet`s.
```

Comment:
Ignore these two sentences about read-only names. I don't know why I added them. It doesn't really come up in the rest of the file. It's mentioned briefly in the "Decision" and "Rationale" section but never really explained. I think I just copied that from the old decision.

**This PR** is supposed to be about the general reference counting used by `Key` & `KeySet`, currently done via `inc`/`dec`(/`del`).

For the read-only names, I'll create a separate decision. It's much simpler for that case, because a re-entrant lock is clearly the only solution.

Reply:
> This PR is supposed to be about the general reference counting used by Key & KeySet, currently done via inc/dec(/del).

We already have reference counting, so the decision must be about what is wrong in what we have now.

Reply:
As noted in https://github.com/ElektraInitiative/libelektra/pull/4805#discussion_r1085399408, this I introduced this decision as a means to re-examine the current refcounting implementation and to explore the different options, since there was no documentation on why the current solution was chosen.

As always "rejecting" the decision is an option. Although in this case for documentation purposes, I would actually still move the decision to "implemented" and just choose the current solution as the "Decision".

So in short: I'm not saying there is necessarily a problem with the current implementation. All I want is that _we_ (not just myself) think about whether another option could be better, _without_ always falling back to "we already have an implementation". I have absolutely no issue, if we decide "the current implementation is good enough" or "another option is better, but not so much that the effort of changing it is worth it". However, before we make that decision, we _have to_ discuss all options equally.